linkerd/app/integration/src/tests/client_policy.rs

```diff
@@ -499,6 +514,268 @@ async fn path_based_routing() {
     proxy.join_servers().await;
 }
 
+#[tokio::test]
+async fn http1_retries() {
+    test_basic_retry(client::http1, server::http1).await
+}
+
+#[tokio::test]
+async fn http2_retries() {
+    test_basic_retry(client::http1, server::http1).await
+}
+
+#[tokio::test]
+async fn http1_doesnt_retry_non_retryable_rule() {
+    test_non_retryable(client::http1, server::http1).await
+}
+
+#[tokio::test]
+async fn http2_doesnt_retry_non_retryable_rule() {
+    test_non_retryable(client::http2, server::http2).await
+}
+
+#[tokio::test]
+async fn http1_doesnt_retry_at_per_request_limit() {
+    test_retry_limit(client::http1, server::http1).await
+}
+
+#[tokio::test]
+async fn http2_doesnt_retry_at_per_request_limit() {
+    test_retry_limit(client::http2, server::http2).await
+}
+
+#[tokio::test]
+async fn http1_retries_post_body() {
+    test_retry_body(client::http1, server::http1, http::Method::POST).await
+}
+
+#[tokio::test]
+async fn http2_retries_post_body() {
+    test_retry_body(client::http2, server::http2, http::Method::POST).await
+}
+
+#[tokio::test]
+async fn http1_retries_put_body() {
+    test_retry_body(client::http1, server::http1, http::Method::PUT).await
+}
+
+#[tokio::test]
+async fn http2_retries_put_body() {
+    test_retry_body(client::http2, server::http2, http::Method::PUT).await
+}
+
+#[tokio::test]
+async fn http1_doesnt_retry_long_body() {
+    test_too_long_retry_body(client::http1, server::http1).await
+}
+
+#[tokio::test]
+async fn http2_doesnt_retry_long_body() {
+    test_too_long_retry_body(client::http2, server::http2).await
+}
+
+// these tests doesn't currently work as we would expect, because backend
+// request timeouts are not turned into synthesized 504s until a layer above the
+// retry layer. instead of being `Ok(Response)`s with a failure status, they're
+// `Err(Error)`s when the retry layer sees them, so they don't get retried.
+// TODO(eliza): figure that out
```

Comment:
this will require either moving the actual `backend_request` timeout *below* the `ClientRescue` layer, so that the timeout can be turned into a 504 (like I initially wanted to do in #2419 before I was talked out of it), or adding a special `errors::respond` layer *just* for turning that timeout into a 504, which feels gross...